DDF-3908 Upgrade to Karaf 4.2.0

brjeter opened the pull request:
#### What does this PR do?

Upgrades from Karaf 4.1.2 to Karaf 4.2.0.
Major Notable Changes in Karaf:

- Slimmed down distribution by reducing the number of libraries built into the distribution under `lib/`. (https://github.com/apache/karaf/commit/e7ca1e209f02d844a481bb5d95a351aba47c8c6c#diff-564886f3f54fbea9414b6d9c0fefc315L193) DDF relies on some of those libraries, primarily for various XML operations, so these needed to be added back in.
- Karaf tightened up how they resolve features, which primarily impacted DDF's feature itests.
- Bumped a minor version in Jetty which has significant changes in their session architecture, which forces some changes in how DDF handles sessions.

#### Who is reviewing it? 
@emmberk @paouelle @tbatie @bakejeyner @blen-desta @ryeats 
#### Select relevant component teams: 
https://github.com/orgs/codice/teams/security
#### Ask 2 committers to review/merge the PR and tag them here.
@coyotesqrl
@lessarderic
@stustison
@clockard
@vinamartin
@bdeining 

#### How should this be tested?
Across Windows and *Nix OS's:
- Manually test basic functionality of DDF.
- Manually modify some properties in custom.system.properties, restart, and verify the properties are updated.
- Drop .cfg and .config files in the /etc directory, and confirm that the configurations are installed successfully.
- Confirm that the bin scripts work as expected.
- Walkthrough the Hardening Guide.
- Confirm that the waitForReady command still works.
- Confirm that installing via each install profile works.
- Confirm that logging still works and is still configurable from org.ops4j.pax.logging.cfg and from the Karaf console. Also, try switching the logging configuration to the xml-style configuration (see comment in org.ops4j.pax.logging.cfg, and confirm that logging still works and is configurable from log4j2.config.xml and from the Karaf console.
- Test that the common console commands are still functional.
- Confirm SSHing into a running instance still works. Confirm SSHing into itests still works.
- Confirm debugging a live system
- Confirm that the Karaf instance can be restarted.
- Confirm that the migration:export command.
- Test that profile console commands still work.
- Test that install with ?dev=true and restart still works 
- Test Automatic Start on System Boot (see Documentation)
- Test using the up and down arrows on the console command line
- Test  log:tail, log:set <level> <logger>, log:get, log:list
- Test that setting the karaf.restart.jvm property to true causes system halt commands to behave as reboots
    - set karaf.restart.jvm to true
    - run the reboot command
    - it should shutdown and restart
#### Any background context you want to provide?
#### What are the relevant tickets?
[DDF-3908](https://codice.atlassian.net/browse/DDF-3908)
#### Screenshots
<!--(if appropriate)-->
#### Checklist:
- [X] Documentation Updated
- [X] Update / Add Unit Tests
- [X] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

Bdthomson commented on distribution/kernel/pom.xml:
```diff
@@ -186,6 +186,38 @@
+                                <artifactItem>
+                                    <groupId>org.apache.servicemix.specs</groupId>
+                                    <artifactId>org.apache.servicemix.specs.activator</artifactId>
+                                    <version>2.9.0</version>
```
Can you pull all the 2.9.0 into a version variable?

emmberk replied:
✏️ Is the servicemix.specs.version tied with the karaf version? Should there be a warning where the version is defined to not update it without upgrading Karaf?

brjeter replied:
I don't think it's a tight coupling like some of the other transitive dependencies, but it _should_ be updated when Karaf is upgraded. Adding that now.

Bdthomson commented on distribution/test/itests/pom.xml:
```diff
@@ -36,7 +36,7 @@
         <pax.url.version>2.4.5</pax.url.version>
         <!-- End pax exam properties -->
-        <cometd.version>3.0.9</cometd.version>
+        <cometd.version>3.1.4</cometd.version>
```
This is used twice, should pull into a variable.

Bdthomson commented on features/kernel/src/main/feature/feature.xml:
```diff
@@ -64,14 +64,15 @@
         <bundle dependency="true" start-level="30">mvn:org.eclipse.jdt.core.compiler/ecj/4.5.1</bundle>
         <bundle start-level="30" dependency="true">mvn:javax.el/javax.el-api/3.0.0</bundle>
-        <!--<bundle start-level="30">mvn:org.ops4j.pax.web/pax-web-jsp/${pax.web.version}</bundle>-->
+        <!--<bundle start-level="30">mvn:org.ops4j.pax.web/pax-web-jsp/${project.version}</bundle>-->
```
Is there a reason we are keeping this commented out bundle?

brjeter replied:
This feature and its contents is copied from pax-web and then modified slightly. That's one of the modifications made, so I think we're just keeping it commented out to be explicit.

Bdthomson commented on platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/security/session/AttributeSharingSessionDataStoreFactory.java:
```diff
@@ -0,0 +1,26 @@
+public class AttributeSharingSessionDataStoreFactory implements SessionDataStoreFactory {
+  @Override
+  public SessionDataStore getSessionDataStore(SessionHandler handler) {
+    AttributeSharingSessionDataStore dataStore = new AttributeSharingSessionDataStore();
```
✏️ Could just return a new dataStore.

Bdthomson commented on platform/platform-paxweb-jettyconfig/src/main/resources/jetty.xml:
```diff
@@ -33,32 +33,44 @@
+        </Arg>
+    </Call>
+
+    <!--<Get name="handler">-->
```
Remove comment?

brjeter replied:
I actually need to uncomment these. It was initially a temporary measure. Assuming @stustison still thinks it's valuable. I might add a mention in the documentation for how to use this `RequestLogHandler` when troubleshooting because I had never heard of it.

brjeter replied:
I decided to put it back in. It was broken, even before the karaf upgrade, but I fixed it up and verified it works.

jhunzik commented on distribution/ddf-common/src/main/resources/etc/org.ops4j.pax.logging.cfg:
```diff
@@ -1,4 +1,4 @@
-################################################################################
+d################################################################################
```
typo?

brjeter replied:
Yes 😅 I fixed that in a another commit.

peterhuffer commented on features/kernel/src/main/feature/feature.xml:
```diff
@@ -64,14 +64,15 @@
         <bundle dependency="true" start-level="30">mvn:org.eclipse.jdt.core.compiler/ecj/4.5.1</bundle>
         <bundle start-level="30" dependency="true">mvn:javax.el/javax.el-api/3.0.0</bundle>
-        <!--<bundle start-level="30">mvn:org.ops4j.pax.web/pax-web-jsp/${pax.web.version}</bundle>-->
+        <!--<bundle start-level="30">mvn:org.ops4j.pax.web/pax-web-jsp/${project.version}</bundle>-->
```
❓ Was the `pax.web.version` -> `project.version` change intentional here?

emmberk replied:
❓ Why is the override for the `pax-http-whiteboard` feature now removed?

brjeter replied:
@clockard pointed out that the `7.2.3` version of this feature does not have the vulnerability, so we don't need to override it anymore. https://github.com/codice/ddf/pull/3635#discussion_r212395785

peterhuffer commented on platform/platform-session-invalidator/src/main/java/org/codice/ddf/platform/session/api/HttpSessionInvalidator.java:
```diff
@@ -13,8 +13,8 @@
 
+import java.util.Map;
 import java.util.function.Function;
-import javax.servlet.http.HttpSession;
```
❓ Can we remove this bundles dependency on the javax.servlet-api since this was removed?

Kjames5269 commented on distribution/ddf-common/src/main/resources/bin/start:
```diff
@@ -23,17 +23,32 @@
-  while [ "${LINK}" ]; do
-	  echo "link: ${LINK}" >&2
-    cd "$(dirname "${LINK}")" || exit 2
+  READLINK_EXISTS=$(command -v readlink &> /dev/null)
```
✏️ Another way of formatting this, it doesn't end up shortening the lines at all... but it makes it makes it _easier_ to maintain
```
realpathHelper() {
OURPWD=${PWD}
    cd "$(dirname "${1}")" || exit 2
    LINK=$($2)
    while [ "${LINK}" ]; do
        echo "link: ${LINK}" >&2
        cd "$(dirname "${LINK}")" || exit 2
        LINK=$($2)
    done
    REALPATH="${PWD}/$(basename "${1}")"
    cd "${OURPWD}" || exit 2
    echo "${REALPATH}"
}
realpathHelperReadlinkExist() {
    ls -l "$(basename "${1}")" | awk -F"-> " '{print $2}'
}
realpathHelperReadlinkExtinct() {
    readlink "$(basename "${1}")"
}

realpath() {
  if [ -z $READLINK_EXIST ]; then
     realpathHelper $1 realpathHelperReadlinkExist
  else
     realpathHelper $1 realpathHelperReadlinkExtinct
}
```

Kjames5269 replied:
👍 Got it didn't know that

Kjames5269 commented on libs/test-common/src/main/java/org/codice/ddf/test/common/features/FeatureUtilities.java:
```diff
@@ -196,7 +151,31 @@ public static void uninstallFeature(FeaturesService featuresService, String feat
+
+  public BundleContext getBundleContext() {
+    Bundle bundle = FrameworkUtil.getBundle(FeatureUtilities.class);
+    if (bundle != null) {
```
✏️ Alternatively this could be written as
```
return (bundle != null) ? bundle.getBundleContext() : null
```
to get rid of some lines

Kjames5269 commented on distribution/ddf-common/src/main/resources-filtered/etc/system.properties:
```diff
@@ -257,14 +257,19 @@ karaf.default.repository=system
+# Sets the maximum size of the local shell command history file. If not set,
+# defaults to 10000 entries.
+#
+# karaf.shell.history.file.maxSize = 10000
```
❓ Is this intended to be commented?

bakejeyner commented on distribution/ddf-common/src/main/resources/bin/start:
```diff
@@ -23,17 +23,32 @@
 # ##############################################################################
```
❓ Is this comment block still valid? Pasted for convenience:
```
# ##############################################################################
# The only change in this script from Karaf's original script is to call ddf
# instead of karaf at the end of the file to gain restart support for critical
# and out of memory errors
# ##############################################################################
```

paouelle replied:
❗️ It should be. Whatever changes was done, this logic should be re-inserted.

bakejeyner commented on distribution/ddf-common/src/main/resources/etc/org.ops4j.pax.logging.cfg:
```diff
@@ -129,6 +129,10 @@ log4j2.appender.osgi-karafConsole.type = PaxOsgi
 log4j2.appender.osgi-karafConsole.name = osgi-karafConsole
 log4j2.appender.osgi-karafConsole.filter = VmLogAppender
 
+# SSHD logger
```
❓ Should this be added to the `log4j.xml` configuration file as well?

emmberk replied:
I don't really understand why this is needed in the first place ([KARAF-5559](https://issues.apache.org/jira/browse/KARAF-5559)) 🤷‍♀️. It might be worth adding a comment here about why these lines are needed.

brjeter replied:
I just added it to suppress the logging as Karaf is now doing. spifly is suppressed for the same reason https://github.com/codice/ddf/blob/211864351d183ed7f423b4f502689b9143619fbf/distribution/ddf-common/src/main/resources/etc/org.ops4j.pax.logging.cfg#L93-L94.  I feel like a comment might be overkill in this instance.

emmberk replied:
I know it was added in Karaf 4.2.1, but I don't understand why. A comment would help prevent someone from removing these seemingly-useless lines in the future. E.g. "The The org.apache.aries.spifly level was set to WARN in Karaf 4.2.1. Do not remove."

paouelle commented on platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/security/session/AttributeSharingSessionDataStore.java:
```diff
@@ -0,0 +1,143 @@
+    return candidates
+        .stream()
+        .filter(
+            c -> {
```
✏️ Easier to read if you create a private method named something like `isSessionNotExpired` and reference that as your filter.

tyler30clemens commented on pom.xml:
```diff
@@ -148,11 +149,11 @@
-        <pax.cdi.version>0.12.0</pax.cdi.version>
+        <jetty.version>9.4.11.v20180605</jetty.version>
+        <pax.cdi.version>1.0.0</pax.cdi.version>
+        <pax.web.version>7.2.3</pax.web.version>
```
❓ Wasn't pax.web and jetty updated relatively recently? Is it already capable of another upgrade?

brjeter replied:
Yes. We tend to try to use the version of pax.web that karaf uses. Technically this is a newer version that avoids the vulnerabilities that we want to avoid, but it's at least within the same major version that Karaf uses.

bakejeyner commented:
Note for Jeter: This permission needs to be added to a policy file:
```
{org.apache.karaf.log.core=[java.io.FilePermission "${ddf.home.perm}\etc\log4j2.xml", "read"]}
```

tbatie commented on libs/test-common/src/main/java/org/codice/ddf/test/common/features/FeatureUtilities.java:
```diff
@@ -196,7 +151,31 @@ public static void uninstallFeature(FeaturesService featuresService, String feat
           .build();
 
-  private static void printInactiveBundles(
+  public void printInactiveBundles() {
```
❓ Can we make these and the following methods private? Especially the service related ones, I don't anyone should be using a different bundle context to load services.

What are your thoughts on creating an interface to prevent exposing methods such as these since we're now using it as a service?

brjeter replied:
I thought creating an interface would be overkill for a test utility. To add onto that, I was under the impression that this would all change when `waitForReady` was redone soon.

tbatie commented on pom.xml:
```diff
@@ -105,9 +106,9 @@
                updated to disable karaf's jetty version.-->
-        <karaf.version>4.1.2</karaf.version>
+        <karaf.version>4.2.0</karaf.version>
         <!-- There is a problem with version 4.1.2 of the karaf-maven-plugin for projects of type feature -->
```
❓ Is this comment still relevant?

tbatie commented on libs/test-common/src/main/java/org/codice/ddf/test/common/features/FeatureUtilities.java:
```diff
@@ -228,13 +207,12 @@ private static void printInactiveBundles(
     boolean ready = false;
-    BundleService bundleService = getService(BundleService.class);
+
+    if (bundleService == null) {
```
❓ Why aren't we using blueprint to inject the bundle service? If we can use blueprint can we remove all these methods related to getting services from the bundle context?

brjeter replied:
I've retested this and investigated some more. This is the exception that's being thrown. It appears that the interaction between pax-exam, blueprint, and Karaf's featureService is breaking down, and the service reference for `BundleService` is not being updated when bundle refreshes are happening even though I've verified the `Import-Service` header for `BundleService` is on this bundle. It's an intermittent issue that depends on timing, though it seems to happen most often with the first feature install in the itests. At this point, I don't think it's worth trying to change.
```
org.osgi.service.blueprint.container.ServiceUnavailableException: The Blueprint container is being or has been destroyed: (objectClass=org.apache.karaf.bundle.core.BundleService)
        at org.apache.aries.blueprint.container.ReferenceRecipe.getService(ReferenceRecipe.java:276)
        at org.apache.aries.blueprint.container.ReferenceRecipe.access$000(ReferenceRecipe.java:57)
        at org.apache.aries.blueprint.container.ReferenceRecipe$ServiceDispatcher.call(ReferenceRecipe.java:341)
        at Proxybea323ac_6212_49bd_a326_9f554304889a.getInfo(Unknown Source)
        at org.codice.ddf.test.common.features.FeatureUtilities.waitForRequiredBundles(FeatureUtilities.java:266)
        at org.codice.ddf.test.common.features.FeatureUtilities.installFeature(FeatureUtilities.java:200)
        at org.codice.ddf.test.common.features.FeatureUtilities.installAndUninstallFeature(FeatureUtilities.java:168)
        at org.codice.ddf.features.camel.test.ITCamelFeatures.installAndUninstallFeature(ITCamelFeatures.java:80)
```

clockard commented on features/camel/src/main/feature/feature.xml:
```diff
@@ -18,7 +18,7 @@
     <repository>mvn:org.apache.karaf.features/standard/${karaf.version}/xml/features</repository>
 
-    <repository>mvn:org.apache.karaf.features/spring/${karaf.version}/xml/features</repository>
+    <repository>mvn:org.apache.karaf.features/spring-legacy/${karaf.version}/xml/features</repository>
```
Curious what it means that it is `legacy` now. Do we need to switch up something to be using a more current spring feature?

brjeter replied:
This has been removed. It was due to our version of camel being old, but now that it's been updated in the baseline this is no longer needed.

clockard commented on features/kernel/src/main/feature/feature.xml:
```diff
@@ -64,14 +64,15 @@
     <!-- Override the pax-web pax-http-whiteboard feature and remove the vulnerable pax-web-jsp bundle-->
     <feature name="pax-http-whiteboard" description="Provide HTTP Whiteboard pattern support" version="${pax.web.version}_1">
-        <feature version="[6.0,6.1)">pax-http</feature>
+        <feature version="[7.2,7.3)">pax-http</feature>
```
This feature override was just so we could remove the vulnerable version of pax-web-jsp. Version 7.2.3 is fine according to sonatypes so we should be able to remove this feature.

emmberk commented on distribution/ddf-common/src/main/resources-filtered/etc/system.properties:
```diff
@@ -257,14 +257,19 @@ karaf.default.repository=system
 # Sets the maximum size of the shell command history. If not set,
 # defaults to 500 entries. Setting to 0 will disable history.
 #
 # karaf.shell.history.maxSize = 0
```
✏️ Support was added in Karaf 4.2.0 for overriding values in `system.properties` (https://github.com/apache/karaf/commit/7fb0a652b27ad15f2106d985681cba0e76b5f75b). Do we want to switch to use `custom.system.properties` instead to keep this file more maintainable?

emmberk replied:
✏️ `karaf.shell.history.maxSize` and `karaf.shell.history.file.maxSize` are commented out in custom.system.properties and can be removed too, right?

✏️ The `#START DDF SETTINGS` comment can be removed.

✏️ The Apache header should be removed at the beginning of this file since it should only contain custom DDF stuff, right?

emmberk commented on pom.xml:
```diff
@@ -105,9 +106,9 @@
                of jetty, the overriding feature will need to be replaced and the blacklist entry
                updated to disable karaf's jetty version.-->
-        <karaf.version>4.1.2</karaf.version>
+        <karaf.version>4.2.0</karaf.version>
```
❓ Why aren't we upgrading to 4.2.1?

brjeter replied:
4.2.1 came out just a couple of days before I got this PR up. I *did* try to go to 4.2.1 and ran into several issues with itests, so holding off for now.

brjeter commented:
I was about to start addressing comments when I discovered a performance issue while doing some final verification on downstream projects. I'm going to look into upgrading to Karaf 4.2.1 to see if that has any effect on the performance. I'll need to dig further if it makes no difference, so the PR is going to have to sit out here while I do that. Just an FYI.

brjeter commented on platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/security/session/AttributeSharingHashSessionIdManager.java:
```diff
@@ -83,37 +78,26 @@
+          getSessionAttributeMapsStream()
+              .filter(e -> subjectName.equals(sessionSubjectExtractor.apply(e.getValue())))
+              .map(e -> e.getKey())
+              .findFirst();
```
No. Before we were holding a mapping of sessionId -> set of sessions (across different contexts). This isn't needed for the purposes of sharing attributes with this new architecture. Also, when we call `invalidate` on the session, it will notify the session handler of all contexts to invalidate their sessions.

brjeter commented on platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/security/session/AttributeSharingHashSessionIdManager.java:
```diff
@@ -83,37 +78,26 @@
-          .map(Reference::get)
-          .filter(Objects::nonNull);
+      if (sessionIdOptional.isPresent()) {
+        idManager.invalidateSession(sessionIdOptional.get());
```
Good catch. I added that check in the `invalidateSession` method. I'm honestly not sure what scenario would cause an invalid session to be triggered to be invalidated, but better to be safe.

Kjames5269 commented on distribution/ddf-common/src/main/resources/etc/blacklisted.properties:
```diff
@@ -1,12 +1,12 @@
+pax-war-tomcat;type=feature;range=[7.0.0,7.2.2]
+pax-jsf-support;type=feature;range=[7.0.0,7.2.2]
+pax-jsf-resources-support;type=feature;range=[7.0.0,7.2.2]
+pax-http-undertow;type=feature;range=[7.0.0,7.2.2]
```
I'm not sure if this affects 4.2 or a later version of karaf but blacklisted.properties has been deprecated.
https://issues.apache.org/jira/browse/KARAF-4181?focusedCommentId=16275591&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16275591

brjeter replied:
👍 I'm looking to replace it with their new way of doing this. https://github.com/codice/ddf/pull/3635#discussion_r215085998

brjeter replied:
DDF ticket created to resolve this: https://codice.atlassian.net/browse/DDF-4146

brjeter commented on catalog/core/catalog-core-directorymonitor/src/test/java/org/codice/ddf/catalog/content/monitor/AbstractComponentTest.java:
```diff
@@ -47,6 +48,8 @@
 
     return options(
+        editConfigurationFilePut(
+            "etc/org.apache.karaf.features.cfg", "serviceRequirements", "disable"),
```
The effect of setting it is that Karaf will not fail to resolve a bundle if there is no other bundle in the context that provides that capability. I forget the exact bundles this started failing on, but there were multiples. Since we are disabling this in production, I thought it made sense to do the same in our itests. https://github.com/codice/ddf/blob/master/distribution/ddf/src/main/filtered-resources/etc/org.apache.karaf.features.cfg#L70
To @emmberk's point, I agree it should go in one of the other `configuration` methods. I had just thrown this in here.

brjeter commented on distribution/ddf-common/src/main/resources-filtered/etc/system.properties:
```diff
@@ -302,6 +307,8 @@ karaf.clean.cache=false
 #
 org.apache.servicemix.specs.debug=false
 org.apache.servicemix.specs.timeout=0
+org.apache.karaf.specs.debug = false
```
Copied from the new system.properties file.

brjeter commented on distribution/ddf-common/src/main/resources/etc/org.apache.karaf.command.acl.shell.cfg:
```diff
@@ -15,4 +15,6 @@ setopt = admin
-while = admin
\ No newline at end of file
+while = admin
+.invoke = admin
```
It _does_ in fact start with a `.`. It's for the ability to invoke any commands in the shell. https://github.com/apache/karaf/blob/989cadc9cab63b224e34bd70ac4ec4fb5ac90f1c/shell/core/src/main/java/org/apache/karaf/shell/impl/console/osgi/secured/SecuredSessionFactoryImpl.java#L63

brjeter commented on distribution/ddf-common/src/main/resources/common-bin.xml:
```diff
@@ -69,7 +69,7 @@
                 </exclude>
                 <exclude>
-                    **/org/eclipse/platform/org.eclipse.osgi/3.11.3/org.eclipse.osgi-3.11.3.jar
+                    **/org/eclipse/platform/org.eclipse.osgi/3.12.50/org.eclipse.osgi-3.12.50.jar
```
I _did_ have to patch version `3.12.50`, but found an installation performance issue that was caused by Equinox. I ended up overriding the version of Equinox to version `3.13.0`, which fixed the installation performance issue and conveniently has the security manager performance change. I did this override in this commit: https://github.com/codice/ddf/pull/3635/commits/211864351d183ed7f423b4f502689b9143619fbf

brjeter commented on platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/security/session/AttributeSharingHashSessionIdManager.java:
```diff
@@ -137,205 +121,77 @@ private void registerSessionManager() {
+  private void invalidateSession(String id) {
+    SessionHandler[] tmp = (SessionHandler[]) _server.getChildHandlersByClass(SessionHandler.class);
+    if (tmp != null && tmp.length > 0) {
+      tmp[0].getSession(id).invalidate();
```
Good catch. I had made the assumption that all session handlers would have a reference to any given session, but that's obviously not true. I just manually tested it and found an issue with the casting on line 150. It's fixed to now iterate through all sessionHandlers until it finds the first matching session.

brjeter commented on platform/security/filter/security-filter-web-sso/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -35,12 +35,12 @@
-    <reference id="webContainer"
-            interface="org.ops4j.pax.web.service.WebContainer"
-            availability="mandatory"/>
+    <!--<reference id="webContainer"-->
```
It should be deleted. Fixed.

brjeter commented on distribution/ddf-common/src/main/resources-filtered/etc/startup.properties:
```diff
@@ -1,16 +1,16 @@
-mvn\:org.apache.felix/org.apache.felix.fileinstall/3.6.0 = 11
-mvn\:org.apache.karaf.features/org.apache.karaf.features.core/4.1.2 = 15
\ No newline at end of file
+mvn\:org.apache.felix/org.apache.felix.fileinstall/3.6.4 = 11
```
Config admin is the one we need to worry about.

brjeter commented on platform/security/filter/security-filter-web-sso/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -35,12 +35,12 @@
-    <reference id="webContainer"
-            interface="org.ops4j.pax.web.service.WebContainer"
-            availability="mandatory"/>
+    <!--<reference id="webContainer"-->
```
Deleted.

brjeter commented on distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java:
```diff
@@ -2874,6 +2874,8 @@ private int getMetacardCacheSize(String sourceId) {
   @Override
   protected Option[] configureCustom() {
-    return options(mavenBundle("ddf.thirdparty", "restito").versionAsInProject());
+    return combineOptions(
```
I believe I put this in there because I was debugging this test directly to fix an issue, which circumvents that.

brjeter commented on features/kernel/src/main/feature/feature.xml:
```diff
@@ -19,8 +19,8 @@
-         The karaf pax.web features (version 6.0.6) are blacklisted-->
+    <!-- Currently karaf 4.2.0 pulls in an older version of pax.web so we need to pull in a newer version here
+                 The karaf pax.web features (version 7.0.0) are blacklisted-->
     <repository>mvn:org.ops4j.pax.web/pax-web-features/${pax.web.version}/xml/features</repository>
```
We were technically pulling in 7.2.3 here. But removing this completely now since the vulnerability is fixed in 7.2.3.

brjeter commented on libs/test-common/src/main/java/org/codice/ddf/test/common/AbstractComponentTest.java:
```diff
@@ -73,7 +74,9 @@
-        getTestBundleOptions().build());
+        getTestBundleOptions().build(),
+        editConfigurationFilePut(
+            "etc/org.apache.karaf.features.cfg", "serviceRequirements", "disable"));
```
I added the default values to the .cfg file in the distro, but I don't think we should change the value at all. I would think we'd want to automatically start bundles with kar files that are deployed. @tbatie do you have any opinion on this?

tbatie replied:
I agree with adding it as a default value in the distro. If we're talking about setting it to true by default I wouldn't do that. Which feature should start up when a kar is being dropped is already defined by the `install` parameter.

brjeter commented on libs/test-common/src/main/java/org/codice/ddf/test/common/AbstractComponentTest.java:
```diff
@@ -73,7 +74,9 @@
-        getTestBundleOptions().build());
+        getTestBundleOptions().build(),
+        editConfigurationFilePut(
+            "etc/org.apache.karaf.features.cfg", "serviceRequirements", "disable"));
```
I don't think any of that has changed between 4.1.2 -> 4.2.0 has it?

emmberk replied:
It wasn't part of the changes, but it's something I noticed while reviewing the changes.

brjeter commented on libs/test-common/src/main/java/org/codice/ddf/test/common/AbstractComponentTest.java:
```diff
@@ -73,7 +74,9 @@
-        getTestBundleOptions().build());
+        getTestBundleOptions().build(),
+        editConfigurationFilePut(
+            "etc/org.apache.karaf.features.cfg", "serviceRequirements", "disable"));
```
DDF ticket created to resolve this: https://codice.atlassian.net/browse/DDF-4146

brjeter commented on distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java:
```diff
@@ -587,8 +587,9 @@ private static Integer findPortNumber(Integer portToTry) {
-                    createSetLogLevelOption("org.codice", logLevel))),
+                    createSetLogLevelOption(
+                        "org.codice.ddf.spatial.ogc.csw.catalog.endpoint", "DEBUG"),
+                    createSetLogLevelOption("org.codice.ddf.itests.common", "DEBUG"))),
```
This was just here for debugging purposes. I will remove this.

brjeter commented on platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/security/session/AttributeSharingSessionDataStore.java:
```diff
@@ -55,7 +55,7 @@ public void updateSessionAttributes(String id, Map<String, Object> sessionAttrib
           id,
           _context.getCanonicalContextPath());
-      synchronized (this) {
+      synchronized (sessionDataMap) {
```
I just realized I need to make this change in the other 2 synchronized blocks in this class.

stustison commented on platform/admin/ui/pom.xml:
```diff
@@ -51,7 +51,7 @@
             <groupId>net.minidev</groupId>
             <artifactId>asm</artifactId>
-            <version>1.0.2</version>
+            <version>${asm.version}</version>
```
Not sure this is correct. I believe the JSON library in use here wants this older version. I recall messing around with this asm stuff during the geotools upgrade.

brjeter replied:
This isn't changing the value of the version; it's just using the existing property. Do you think this should be more explicit to avoid being upgraded when the property value changes?

stustison commented on platform/pom.xml:
```diff
@@ -168,7 +168,7 @@
                 <groupId>net.minidev</groupId>
                 <artifactId>asm</artifactId>
-                <version>1.0.2</version>
+                <version>${asm.version}</version>
```
might be the same issue on this asm version as I mentioned up above

emmberk commented on distribution/docs/src/main/resources/content/_managing/_configuring/configuring-cas.adoc:
```diff
@@ -91,8 +91,8 @@ active state. Additionally, it should respond to LDAP requests on the default po
 folder, making sure they overwrite the existing serverKeystore.jks and serverTruststore.jks files.
-. Open the `<DISTRIBUTION>/etc/system.properties file` and make sure the keystore passwords are set
-correctly.
+. Open the `<DISTRIBUTION>/etc/custom.system.properties file` and make sure the keystore passwords
```
✏️ @ricklarsen Shouldn't these paths be `${home_directory}/etc` instead of `<DISTRIBUTION>/etc`?

ricklarsen replied:
yes

emmberk commented on distribution/docs/src/main/resources/content/_managing/_configuring/configuring-sts.adoc:
```diff
@@ -76,7 +76,7 @@ Be sure to limit the hosts that are allowed to connect to the STS:
 * *{hardening-step}*
 
-* Open the `<${branding}_HOME>/etc/system.properties` file.
+* Open the `<${branding}_HOME>/etc/custom.system.properties` file.
```
✏️ `${home_directory}/etc`

ricklarsen replied:
yes

emmberk commented on distribution/docs/src/main/resources/content/_managing/_running/solr-service.adoc:
```diff
@@ -14,7 +14,7 @@ These instructions are for configuring Solr as a service managed by the operatin
 
 . If ${branding} is running, stop it.
-. Edit `<${branding}_HOME>/etc/system.properties` and set `start.solr=false`.
+. Edit `<${branding}_HOME>/etc/custom.system.properties` and set `start.solr=false`.
```
✏️ `${home_directory}/etc`

ricklarsen replied:
yes

emmberk commented on distribution/docs/src/main/resources/content/_managing/_running/solr-service.adoc:
```diff
@@ -52,7 +52,7 @@ These instructions are for unix operating systems running the systemd
 
 . If ${branding} is running, stop it.
-. Edit `<${branding}_HOME>/etc/system.properties` and set `start.solr=false`.
+. Edit `<${branding}_HOME>/etc/custom.system.properties` and set `start.solr=false`.
```
✏️ `${home_directory}/etc`

ricklarsen replied:
yes

emmberk commented on platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/security/session/AttributeSharingSessionDataStore.java:
```diff
@@ -0,0 +1,141 @@
+  public void updateSessionAttributes(String id, Map<String, Object> sessionAttributes) {
+    SessionData sessionData = sessionDataMap.get(id);
+    if (sessionData != null && !sessionData.getAllAttributes().equals(sessionAttributes)) {
+      LOGGER.debug(
```
✏️ `trace` because this is normal program execution. See https://codice.atlassian.net/wiki/spaces/DDF/pages/7045131/Log+Levels

brjeter replied:
I definitely remember changing this. I must've undone it somehow before committing 😅

emmberk commented on libs/test-common/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -0,0 +1,18 @@
+<blueprint xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0">
+  <service interface="org.codice.ddf.test.common.features.FeatureUtilities">
+      <bean class="org.codice.ddf.test.common.features.FeatureUtilities"/>
+    </service>
```
✏️ Formatting still looks off here

brjeter replied:
I remember changing this... Fixing.

coyotesqrl commented on platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/security/session/AttributeSharingSessionDataStore.java:
```diff
@@ -127,15 +131,18 @@ public SessionData load(String id) {
   private boolean isExpired(String candidateId, long now) {
-    SessionData sessionData = sessionDataMap.get(candidateId);
+    SessionData sessionData;
+    synchronized (sessionDataMap) {
```
❗️ The `exists()` and `load()` methods should also sync on the map.

paouelle commented on platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/security/session/AttributeSharingSessionDataStore.java:
```diff
@@ -49,9 +49,13 @@
   public void updateSessionAttributes(String id, Map<String, Object> sessionAttributes) {
-    SessionData sessionData = sessionDataMap.get(id);
+    SessionData sessionData;
+    synchronized (sessionDataMap) {
```
You go to put both `get()` and `put()` underneath the same synchronized block.

paouelle commented on platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/security/session/AttributeSharingSessionDataStore.java:
```diff
@@ -49,17 +49,14 @@
+        LOGGER.trace(
+            "Storing new attributes for session {} at context {}",
+            id,
+            _context.getCanonicalContextPath());
```
`newSessionData() below ... should it not cache the session being created?

brjeter replied:
It hasn't created a new Map. Rather, it's just clearing and putting new elements in the already cached Map.

brjeter replied:
I misunderstood the question in the above response, but @paouelle and I hashed it out offline. The `Map` created from `newSessionData` is later cached when Jetty's `SessionHandler` calls the `doStore` method.

paouelle commented:
Some of the steps in the Hero section are extracted from the previous Karaf upgrade where we were verifying Karaf bug fixes we were waiting for. I am not sure we should keep verifying those over and over again.

bakejeyner commented:
Windows Hero ✔️

blen-desta commented:
✅ Hero'd successfully on mac

bdeining commented:
build now

brjeter commented:
build now (codice-parent artifacts from https://github.com/codice/codice-parent/pull/6 are now deployed)

blen-desta commented:
Waiting for https://github.com/codice/thirdparty/ to get deployed before starting a new build.

bdeining commented:
@blen-desta thirdparty tika is deployed, this CI failed for some other reason

blen-desta commented:
build now
